src/main/flight/mixer.c

```diff
@@ -526,6 +544,13 @@ void writeMotors(void)
         }
 #endif
         for (int i = 0; i < motorCount; i++) {
+            if (revFlags[i]) {
+                if (motor[i] == disarmMotorOutput) {
```

Comment:
disarmMotorOutput is set to motorConfig()->mincommand near line 405.
If disarmed the reverse servo should travel to the disarmed position. Did i have a lapse of thought?

Reply:
@mikeller 
Slept a night over it. For helicopter and variable pitch use `mincommand` and `minthrottle` should be the same so there is no need for this lines. 
But what if someone sets it to something different for whatever reason? How to handle it? 

What about `stopMotors(void)` and `stopPwmAllMotors(void)`? How to handle?

Reply:
IMO whole concept of reversing motors (to help with servos implementation) is problematic.
For motor reversion makes sense only in 3D mode and should be equivalent to reversing motor direction (isn't is already supported in BF?).
Servos should reuse existing servo implementation (possibly resurrecting removed code / porting code from iNav) and handle servo-specific details in separate layer.

Reply:
@ledvinap: I don't think 'reversing motors' is the aim here - this is intended to drive the swash plate servos on a helicopter from the mixer output. What this does is essentially apply the `smix reverse` functionality to motors.

@Woga65: I think @ledvinap has got a point, and this is most likely better off in the servo mixer functionality - sorry to have given you bad advice on this. Looking at
https://github.com/betaflight/betaflight/blob/master/docs/Mixer.md#custom-servo-mixing, if you defined a set of rules for each of 'Stabilized ROLL', 'Stabilized PITCH', 'Stabilized YAW', and 'THROTTLE' with the respective rate for each of your servos you should be able to define the equivalent of what the motor mixer does. https://github.com/betaflight/betaflight/blob/master/docs/Mixer.md#example-1-a-kk20-wired-motor-setup can serve as an example, except that all of the components of each line will need to use a line of their own. This should make it feasible to use the firmware without code change.

Reply:
@mikeller: I don't think so. The servo handling in Betaflight is limited in a way that makes it nearly impossible to do a proper mixing. You can see it in the code: Every vehicle different from a classic multi-rotor needs some custom code wasting rare memory space. the same applies to iNav. 
That is why all the attempts to implement a CCPM mixer failed. Yes, it can be done but it would be much too limited for practical use.

Maybe the servo support other than `servo_tilt` / `channel forwarding` can be even dropped if using the motor mixer. 

I had a further look into the code as well as some testing on an OMNIBUSF4SD. When proper setup the difference between `mincommand` and `minthrottle`  can be safely ignored.

Reply:
@woga65: 

> The servo handling in Betaflight is limited in a way that makes it nearly impossible to do a proper mixing.

Can you explain what you think is missing from the facilities described in https://github.com/betaflight/betaflight/blob/master/docs/Mixer.md#custom-servo-mixing to make your use case work? To my understanding this can do everything that the motor mixer can do as well, plus it can do the reversal that you are looking for.

Reply:
For example,
- `servo_tilt` does not work if using a servo related mixer
- the throttle channel (pitch if setting up a heli or v-pitch quad) is limited to pwm-output 1

Remember, all of the swash-plate servos need to respond to the throttle channel and the ESC(s) need to operate independently from the mixer. Also the vehicle need to remain armed even if the ESC(s) are powered down.

Reply:
This is why the remaining fragments of the CCPM-120 mixer use an AUX channel for the collective pitch.

Reply:
Quote:

0 | GIMBAL PITCH
-- | --
1 | GIMBAL ROLL
2 | ELEVATOR / SINGLECOPTER_4
3 | FLAPPERON 1 (LEFT) / SINGLECOPTER_1
4 | FLAPPERON 2 (RIGHT) / BICOPTER_LEFT / DUALCOPTER_LEFT / SINGLECOPTER_2
5 | RUDDER / BICOPTER_RIGHT / DUALCOPTER_RIGHT / SINGLECOPTER_3
6 | THROTTLE (Based ONLY on the first motor output)
7 | FLAPS

Only some <servo> channels are connected to output, based on mode. For custom modes:

    RUDDER for CUSTOM_TRI
    ELEVATOR ... FLAPS for CUSTOM_AIRPLANE
    no servos for CUSTOM

Reply:
@woga65:

> `servo_tilt` does not work if using a servo related mixer

I do not know how you get to this conclusion - `SERVO_TILT` uses attitude information to control the gimbal servos. This is completely independent from what mixer is used to attain the attitude: https://github.com/betaflight/betaflight/blob/0c8c671cde69dd96a63e5f75d817a83e27195866/src/main/flight/servos.c#L509-L513

> the throttle channel (pitch if setting up a heli or v-pitch quad) is limited to pwm-output 1

That is simply incorrect. The throttle channel is not limited in any way.

> ELEVATOR ... FLAPS for CUSTOM_AIRPLANE

So 6 servo channels are not enough for what you want to achieve? In this case I doubt that you will be able to achieve what you want to do with the motor mixer either.

Reply:
IMO correct way to handle this is to fix servo mixer.

Either make it fully generic (allowing fully customized motor and servo mixer) or implement heli-specific configuration ... 

Hacking existing code is fine for experimentation, but it is not manageable in long term (sadly a lot of heuristic/quick to implement code is in BF)

Reply:
What does this mean?

6 | THROTTLE (Based ONLY on the first motor output)
-- | --

Um. For me `servo_tilt` did not work if using for example the `CUSTOM_TRI` mixer. 
I need to have a closer look at it.

So a configuration like the following will work?

Servo 1 | +roll | -pitch | +collective | 0 
Servo 2 | -roll | -pitch | +collective | 0 
Servo 3 | 0    | +pitch | +collective | 0 
Servo 4 | 0    | 0        | 0              | +yaw 
ESC 1 | AUX

Collective = throttle

Reply:
@mikeller / @ledvinap 
Since the documentation does not cover it explicitly and there is a lot of misinformation  on the internet, please allow two more questions.

- Are the assignments of servos / slots and its functions to the output pins hard coded or may I move them freely using the assign command?
- When using the mixer `CUSTOM_AIRPLANE` and not using motors, may i use the outputs no 1+2 for `smix` ?

Reply:
Thank you very much for making this clear!
The servo mixer works for helicopters!
Apart from the `auxpGain` -thing no additional code to make it finally work.

The settings for the target `OMNIBUSF4SD` are:

resource SERVO 1 A03
resource SERVO 2 A02
resource SERVO 3 A01
resource SERVO 4 A08
resource SERVO 5 B00
resource SERVO 6 B01

smix 0 3 0 100 0 0 100 0
smix 1 3 1 100 0 0 100 0
smix 2 3 7 100 0 0 100 0
smix 3 4 0 100 0 0 100 0
smix 4 4 1 100 0 0 100 0
smix 5 4 7 100 0 0 100 0
smix 6 2 1 100 0 0 100 0
smix 7 2 7 100 0 0 100 0
smix 8 5 2 100 0 0 100 0
smix 9 0 10 100 0 0 100 0
smix 10 1 11 100 0 0 100 0 

I will close this request and make a new one including just the PID adjustment-channel.

Reply:
@woga65: Yes, this is pretty much how I would have expected this to have to be done - build up the mix for every output by using one line per input going into it. And yes, it needs raw throttle, 'stabilised THROTTLE' is an odd concept that takes what would be the motor output of motor 0 of the motor mixer as the 'throttle' input - the only way if this is ever going to be useful if you configure a motor mixer with only 1 output, and probably even then it will only add actual value in the case of altitude hold (which does not exist any more) being used.

Not sure that you need to add `smix` lines for servos 0 and 1 (the `SERVO_TILT` / gimbal servos) - these seem to be pretty much hard wired to work as gimbal servos if a corresponding feature is enabled, and are ignored by all other mixers. (It would probably even make sense to separate these two servos out better, at least in the documentation.)

As you point it out, the documentation is not very good, and in some places outdated. I think part of this is caused by the fact that the servo mixer has a 'swiss army knive' approach that makes it flexible and powerful, but also makes it hard to understand, so even in the best of cases the documentation will not make it easy to grasp.
Not sure if there is much that can be changed to make the documentation better, but if you see potential, can you please open a pull request for it? I'd also be very interested to get you to document your work for the helicopter mixer in this pull request once you have finalised it - I think this is a good example of what can be done with the servo mixer.

Reply:
@mikeller 
Yes, to lines
smix 9 0 10 100 0 0 100 0
smix 10 1 11 100 0 0 100 0
are not needed. Either `servo_tilt` or `channel_forwarding` can do the job.

Anyway, perhaps I had been glad too early! 
To make it work a simple servo reverse might not be sufficient. Also the direction of the stabilization need to be configurable per output like it is in the motor mixer.
What code should i have a look?

Reply:
@woga65: As I said, it is a Swiss army knive by design, so I almost expect it to be applicable to your case by design. If the `smix ...` does not give you enough capability to reverse, you can also entirely reverse the output of a servo with `smix reverse ...` on top of that.

Reply:
@mikeller Things would be much easier if servos were internal reversible in a standardized way. Some servos are programmable, some are not, often the programming interface is not available...

Reply:
It is done! Just a few lines of custom code needed to adjust the PID according to the head speed. https://youtu.be/xKXoeUIHnWs
I will make a new pull request with just these few lines.

Reply:
@woga65: Instead of opening a new pull request, can I recommend that you just repurpose this one? You can pretty much completely rewrite your branch, and the pull request will update accordingly. This way the discussion and the outcome stay in the same place.

Reply:
@mikeller Thank you for the advice! So i will pull the original ßb code, re-apply the changes regarding the adjustment channel, then commit and push.